javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll

```diff
@@ -211,89 +212,333 @@ module TaintTracking {
+        any(SharedTaintStep step).stringManipulationStep(pred, succ)
+      }
+
+      /**
+       *  Holds if `pred` &rarr; `succ` should be considered a taint-propagating
+       * data flow edge through data serialization, such as `JSON.stringify`.
+       */
+      cached
+      predicate serializeStep(DataFlow::Node pred, DataFlow::Node succ) {
+        any(SharedTaintStep step).serializeStep(pred, succ)
+      }
+
+      /**
+       * Holds if `pred` &rarr; `succ` should be considered a taint-propagating
+       * data flow edge through data deserialization, such as `JSON.parse`.
+       */
+      cached
+      predicate deserializeStep(DataFlow::Node pred, DataFlow::Node succ) {
+        any(SharedTaintStep step).deserializeStep(pred, succ)
+      }
+
+      /**
+       * Holds if `pred` &rarr; `succ` should be considered a taint-propagating
+       * data flow edge through a promise.
+       *
+       * These steps consider a promise object to tainted if it can resolve to
+       * a tainted value.
+       */
+      cached
+      predicate promiseStep(DataFlow::Node pred, DataFlow::Node succ) {
+        any(SharedTaintStep step).promiseStep(pred, succ)
+      }
     }
   }
 
+  import Cached::Public
+
   /**
-   * Holds if there is taint propagation through the heap from `pred` to `succ`.
+   * Holds if `pred -> succ` is a taint propagating data flow edge through a string operation.
    */
-  private predicate heapStep(DataFlow::Node pred, DataFlow::Node succ) {
-    exists(Expr e, Expr f | e = succ.asExpr() and f = pred.asExpr() |
-      exists(Property prop | e.(ObjectExpr).getAProperty() = prop |
-        prop.isComputed() and f = prop.getNameExpr()
+  pragma[inline]
+  predicate stringStep(DataFlow::Node pred, DataFlow::Node succ) {
+    stringConcatenationStep(pred, succ) or
+    stringManipulationStep(pred, succ)
+  }
+
+  /**
+   * Holds if `pred -> succ` is an edge used by all taint-tracking configurations.
+   */
+  predicate sharedTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
+    Cached::legacyAdditionalTaintStep(pred, succ) or
+    Cached::genericStep(pred, succ) or
+    Cached::heuristicStep(pred, succ) or
+    uriStep(pred, succ) or
+    persistentStorageStep(pred, succ) or
+    heapStep(pred, succ) or
+    arrayStep(pred, succ) or
+    viewComponentStep(pred, succ) or
+    stringConcatenationStep(pred, succ) or
+    stringManipulationStep(pred, succ) or
+    serializeStep(pred, succ) or
+    deserializeStep(pred, succ) or
+    promiseStep(pred, succ)
+  }
+
+  /**
+   * A taint-propagating data flow edge that should be added to all taint tracking
+   * configurations in addition to standard data flow edges.
+   *
+   * Note: For performance reasons, all subclasses of this class should be part
+   * of the standard library. Override `Configuration::isAdditionalTaintStep`
+   * for analysis-specific taint steps.
+   */
+  abstract class AdditionalTaintStep extends DataFlow::Node {
```

Comment:
Could we mark this as deprecated, or at least add some prominent text about `SharedTaintSted`? Otherwise, the users will keep using it.
(I suppose the reason that this isn't deprecated is that we use it ourselves in `legacyAdditionalTaintStep`, it would be nice if we could juggle our way out of that)

Reply:
Oh I thought it actually was deprecated. It's properly deprecated now.

I've also updated the tutorial. In the process of doing that, I introduced a forward-compatible version of `DataFlow::SharedFlowStep` as one paragraph simply couldn't be updated without distracting the reader with the details of how these two classes differ. Actually migrating to `SharedFlowStep` is for a future PR.